components/eppp_link/eppp_link.c

```diff
@@ -0,0 +1,802 @@
+/*
+ * SPDX-FileCopyrightText: 2019-2024 Espressif Systems (Shanghai) CO LTD
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <string.h>
+#include <stdint.h>
+#include "sdkconfig.h"
+#include "esp_log.h"
+#include "esp_netif.h"
+#include "esp_check.h"
+#include "esp_event.h"
+#include "esp_netif_ppp.h"
+#include "eppp_link.h"
+
+#if CONFIG_EPPP_LINK_DEVICE_SPI
+#include "driver/spi_master.h"
+#include "driver/spi_slave.h"
+#include "driver/gpio.h"
+#include "esp_timer.h"
+#elif CONFIG_EPPP_LINK_DEVICE_UART
+#include "driver/uart.h"
+#endif
+
+static const int GOT_IPV4 = BIT0;
+static const int CONNECTION_FAILED = BIT1;
+#define CONNECT_BITS (GOT_IPV4|CONNECTION_FAILED)
+
+static EventGroupHandle_t s_event_group = NULL;
+static const char *TAG = "eppp_link";
+static int s_retry_num = 0;
+static int s_eppp_netif_count = 0; // used as a suffix for the netif key
+
+struct eppp_handle {
+#if CONFIG_EPPP_LINK_DEVICE_SPI
+    QueueHandle_t out_queue;
+    QueueHandle_t ready_semaphore;
+    spi_device_handle_t spi_device;
+    spi_host_device_t spi_host;
+    int gpio_intr;
+#elif CONFIG_EPPP_LINK_DEVICE_UART
+    QueueHandle_t uart_event_queue;
+    uart_port_t uart_port;
+#endif
+    esp_netif_t *netif;
+    enum eppp_type role;
+    bool stop;
+    bool exited;
+    bool netif_stop;
+};
+
+struct packet {
+    size_t len;
+    uint8_t *data;
+};
+
+static esp_err_t transmit(void *h, void *buffer, size_t len)
+{
+    struct eppp_handle *handle = h;
+#if CONFIG_EPPP_LINK_DEVICE_SPI
+#define MAX_PAYLOAD 1600
+    struct packet buf = { };
+
+    uint8_t *current_buffer = buffer;
+    size_t remaining = len;
+    do {
```

Comment:
This part is only used for SPI transport, which sends/receives frames of a certain maximum size per iteration. It is convenient to produce data for one iteration of the sending loop.
(Note that with the current settings we always queue one buffer per transmit function -- per LWIP MTU settings, but this allows to reduce the max size of SPI frames, e.g. if there're multiple devices on the same bus..?)

Reply:
My point here is to reduce the number of allocations and let the sending loop to control its iterations. We will still keep the same amount of iterations to SPI, but will reduce the allocations to the cost of moving the split of the data to the sending loop iteration control.

Reply:
Yes, could reduce number of allocation at the expense of more code. I'll check if I can rework to move this to the task. At this moment I'd tend to add it to the comment as `TODO` (for possible future improvements), since this is a rare scenario with specific config only.

Reply:
Added a TODO, as mentioned above.